model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleModelCriteriaBuilder.java

```diff
@@ -0,0 +1,140 @@
+import org.keycloak.models.map.storage.jpa.JpaModelCriteriaBuilder;
+import org.keycloak.models.map.storage.jpa.role.entity.JpaRoleEntity;
+import org.keycloak.storage.SearchableModelField;
+
+public class JpaRoleModelCriteriaBuilder extends JpaModelCriteriaBuilder implements ModelCriteriaBuilder<RoleModel, JpaRoleModelCriteriaBuilder> {
+
+    private BiFunction<CriteriaBuilder, Root<JpaRoleEntity>, Predicate> predicateFunc = null;
+
+    public JpaRoleModelCriteriaBuilder() {
+    }
+
+    private JpaRoleModelCriteriaBuilder(BiFunction<CriteriaBuilder, Root<JpaRoleEntity>, Predicate> predicateFunc) {
+        this.predicateFunc = predicateFunc;
+    }
+
+    private String getFieldName(SearchableModelField<? super RoleModel> modelField) {
+        return "f" + modelField.getName().substring(0, 1).toUpperCase() + modelField.getName().substring(1);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public JpaRoleModelCriteriaBuilder compare(SearchableModelField<? super RoleModel> modelField, Operator op, Object... value) {
+        switch (op) {
+            case EQ:
+                if (modelField.equals(SearchableFields.REALM_ID) ||
+                    modelField.equals(SearchableFields.CLIENT_ID)) {
+
+                    validateValue(value, modelField, op, String.class);
+
+                    return new JpaRoleModelCriteriaBuilder((cb, root) -> 
+                        cb.equal(
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal(getFieldName(modelField))), 
+                            cb.literal(convertToJson(value[0]))
+                        )
+                    );
+                } else if (modelField.equals(SearchableFields.NAME)) {
+                    validateValue(value, modelField, op, String.class);
+
+                    return new JpaRoleModelCriteriaBuilder((cb, root) -> 
+                        cb.equal(root.get(modelField.getName()), value[0])
+                    );
+                } else if (modelField.equals(SearchableFields.IS_CLIENT_ROLE)) {
+                    validateValue(value, modelField, op, Boolean.class);
+
+                    return new JpaRoleModelCriteriaBuilder((cb, root) -> 
+                        cb.equal(
+                            cb.function("->", JsonbType.class, root.get("metadata"), cb.literal("fClientRole")), 
+                            cb.literal(convertToJson(value[0]))
+                        )
+                    );
+                } else {
+                    throw new CriterionNotSupportedException(modelField, op);
+                }
+            case IN:
+                if (modelField.equals(SearchableFields.ID)) {
+                    validateValue(value, modelField, op, String.class);
+
+                    return new JpaRoleModelCriteriaBuilder((cb, root) -> {
+                        In<String> in = cb.in(root.get("id"));
+                        ((Stream<String>) value[0]).forEach(id -> in.value(id));
```

Comment:
This cast is dangerous. `value[0]` is expected to be `String` in line 84, and generally can be any `Collection` or `Stream`

Reply:
Thank you